sled-diagnostics/src/logs.rs

```diff
@@ -0,0 +1,245 @@
+use std::collections::{BTreeMap, BTreeSet, HashMap};
+
+use camino::Utf8PathBuf;
+use schemars::{
+    JsonSchema,
+    gen::SchemaGenerator,
+    schema::{Schema, SchemaObject},
+};
+use serde::{Deserialize, Serialize};
+use thiserror::Error;
+
+pub type Zone = String;
+pub type Service = BTreeMap<String, Vec<Utf8PathBuf>>;
+
+#[derive(Serialize, Deserialize, Default, JsonSchema)]
+pub struct SledDiagnosticsLogs {
+    #[schemars(schema_with = "path_schema")]
+    logs: BTreeMap<Zone, Service>,
+}
+
+impl SledDiagnosticsLogs {
+    pub fn logs(&self) -> &BTreeMap<Zone, Service> {
+        &self.logs
+    }
+}
+
+#[derive(Debug, Error)]
+pub enum SledDiagnosticsLogError {
+    #[error("Failed to find log paths: {0}")]
+    ZonePaths(anyhow::Error),
+}
+
+// Used for schemars to be able to be used with camino:
+// See https://github.com/camino-rs/camino/issues/91#issuecomment-2027908513
+fn path_schema(gen: &mut SchemaGenerator) -> Schema {
+    let mut schema: SchemaObject = <String>::json_schema(gen).into();
+    schema.format = Some("Utf8PathBuf".to_owned());
+    schema.into()
+}
+
+/// Provides all of the valid directores logs can be found in for this sled.
+pub fn valid_log_dirs() -> Result<BTreeSet<Utf8PathBuf>, SledDiagnosticsLogError>
+{
+    let zones =
+        oxlog::Zones::load().map_err(SledDiagnosticsLogError::ZonePaths)?;
+    Ok(zones
+        .zones
+        .into_values()
+        .flat_map(|entry| {
+            let mut paths = Vec::new();
+            paths.push(entry.primary);
+            paths.extend(entry.debug);
+            paths.extend(entry.extra.into_iter().map(|e| e.1));
+            paths
+        })
+        .collect())
+}
+
+/// Find the relevant logs that should be returned as apart of a Support Bundle.
+pub fn find_logs() -> Result<SledDiagnosticsLogs, SledDiagnosticsLogError> {
+    let zones =
+        oxlog::Zones::load().map_err(SledDiagnosticsLogError::ZonePaths)?;
+
+    let mut result = SledDiagnosticsLogs { logs: BTreeMap::new() };
+    for zone in zones.zones.keys() {
+        let filter = oxlog::Filter {
+            current: true,
+            archived: true,
+            extra: true,
+            // avoid having to call `lstat` on each log file
+            show_empty: true,
+        };
+
+        let mut logs = Service::new();
+        for (svc, mut svclogs) in zones.zone_logs(&zone, filter) {
+            // Current
+            if let Some(current) = svclogs.current {
+                logs.insert(svc.clone(), vec![current.path]);
+            }
+
+            // Archived
+            logs.entry(svc.clone()).or_default().extend(
+                svclogs.archived.into_iter().rev().take(5).map(|s| s.path),
```

Comment:
What's the `rev().take(5)` for?

Reply:
Recall that we don't yet have any log rotation cleanup logic and environments like dogfood contain many log files, so we are taking the most recent 5 log rotations.

Reply:
Aren't we using https://github.com/oxidecomputer/omicron/blob/0f81eb138b85f9fd1444a4cdedf86ea655b7670c/smf/logadm/logadm.conf ? I'm trying to map that to https://illumos.org/man/8/logadm , but doesn't that use the `-C` argument, which limits the total count of logs?

Reply:
(I'm not disagreeing with you, just trying to better grok which logs are unbounded)